### ISSUE TITLE
lint for integer overflow, especially in release mode

### ISSUE THREAD

**Issue by bingmatv** [rust-lang/rust-clippy#12503]
### What it does

Rust panics for integer overflow in debug mode, but it will wrap around any integers in release mode, aka, when `--release` parameter is added for cargo.

### Example

`fn main() {
    let mut s = String::new();
    std::io::stdin().read_line(&mut s).unwrap();
    let s = (s.trim().parse::<u8>().unwrap()) * 3;
}`
It accepts user input then multiply the accepted integer by 3, may wrap around when multiply by 3.

**Comment by taiki-e** [comment]
This seems to be covered by [arithmetic_side_effects](https://rust-lang.github.io/rust-clippy/master/index.html#/arithmetic_side_effects)?

**Comment by bingmatv** [comment]
> This seems to be covered by [arithmetic_side_effects](https://rust-lang.github.io/rust-clippy/master/index.html#/arithmetic_side_effects)?

`cargo clippy` only showed:
warning: unused variable: `s`
 --> src/main.rs:4:9
  |
4 |     let s = (s.trim().parse::<u8>().unwrap()) * 3;
  |         ^ help: if this is intentional, prefix it with an underscore: `_s`
Only warned about unused variable.

**Comment by taiki-e** [comment]
arithmetic_side_effects is in the `restriction` lint group, which is allowed by default, so you have to explicitly enable it.

```rs
#![warn(clippy::arithmetic_side_effects)] // <-------------
#![allow(dead_code, unused_variables)]

fn main() {
    let mut s = String::new();
    std::io::stdin().read_line(&mut s).unwrap();
    let s = (s.trim().parse::<u8>().unwrap()) * 3;
}
```

```
warning: arithmetic operation that can potentially result in unexpected side-effects
 --> src/main.rs:7:13
  |
7 |     let s = (s.trim().parse::<u8>().unwrap()) * 3;
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#arithmetic_side_effects
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![warn(clippy::arithmetic_side_effects)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7c8e8fb4399dba9f3230af8061250c1c)

**Comment by ijackson** [comment]
Unfortunately `arithmetic_side_effects` is quite poor:

 * It can't manage to realise even the simplest cases where overflow is statically impossible because of constant parameters (#11266)
 * It triggers on user-defined types (so wrapping up a primitive integer in a panic-free API gives lots of FPs) (#11220)
 * It doesn't notice if you call eg `.add_assign` (#12318)

But probably it doesn't make sense to use this ticket as a tracking issue for problems with this lint; it shoud be closed.
